crates/sui-types/src/dynamic_field/visitor.rs

```diff
@@ -46,6 +53,22 @@ impl FieldVisitor {
     }
 }
 
+impl<'b, 'l> Field<'b, 'l> {
+    /// If this field is a dynamic field, returns its value's type. If it is a dynamic object
+    /// field, it returns the ID of the object the value points to (which must be fetched to
+    /// extract its type).
+    pub fn value_type(&self) -> Result<TypeTag, Error> {
+        match self.kind {
+            DynamicFieldType::DynamicField => Ok(TypeTag::from(self.value_layout)),
+            DynamicFieldType::DynamicObject => {
+                let id: ObjectID = bcs::from_bytes(&self.value_bytes)
+                    .map_err(|_| Error::NotADynamicObjectField)?;
+                Err(Error::DynamicObjectField(id))
+            }
+        }
+    }
```

Comment:
It feels strange to me to use the err variant to communicate the dynamic object id. Would it be clearer to have this separated into two distinct methods, or do you think its generally better to have this merged?

Reply:
Yeah it is a bit strange -- I'm in two minds about it. Maybe it'll help to explain how I ended up here, and then a better solution might emerge:

- I started with two functions: `value_type` and `value_object_id`, both returning `Result<Option<_>, Error>`,
- but found they were mutually exclusive and when they were used, it was usually done by matching on the kind, and then calling one or the other, which then implies an unwrap or a redundant test that the relevant value does in fact exist.
- So, I brought them together and I modeled it as a function where you ask for the value type which you either have access to, or you don't because the value is a dynamic object field.

Going over this, I'm thinking the most sane approach would be to just suck it up and introduce a new type, like:

```rust
pub enum ValueMetadata {
    DynamicField(TypeTag),
    DynamicObjectField(ObjectID),
}
```

and return that instead:

```rust
pub fn value_metadata(&self) -> Result<ValueMetadata, Error> {
    /* ... */
}
```

Reply:
Yep i'd agree with you that the new type (essentially an `Either`) would communicate a much clearer api